tests/test_database.py

```diff
@@ -0,0 +1,105 @@
+
+
+@pytest.fixture(scope='function')
+def db_sa(db):
+    # To make it easier to test, we keep the test restricted to firebase_tests
+    # Because of the current mutations on calls, we return it as a function.
+    name = 'test_%05d' % random.randint(0, 99999)
+    yield lambda: db().child(name)
+
+
+@contextmanager
+def make_stream(db, cbk):
+    s = db.stream(cbk)
+    try:
+        yield s
+    finally:
+        s.close()
+
+
+@contextmanager
+def make_append_stream(db):
+    l = []
+
+    def cbk(event):
+        l.append(event)
+
+    with make_stream(db, cbk) as s:
+        yield s, l
+
+
+class TestSimpleGetAndPut:
+    def test_simple_get(self, db_sa):
+        assert db_sa().get().val() is None
+
+    def test_put_succeed(self, db_sa):
+        assert db_sa().set(True)
+
+    def test_put_then_get_keeps_value(self, db_sa):
+        db_sa().set("some_value")
+        assert db_sa().get().val() == "some_value"
+
+    def test_put_dictionary(self, db_sa):
+        v = dict(a=1, b="2", c=dict(x=3.1, y="3.2"))
+        db_sa().set(v)
+
+        assert db_sa().get().val() == v
+
+    @pytest.mark.skip
+    def test_put_deeper_dictionnary(self, db_sa):
+        v = {'1': {'11': {'111': 42}}}
+        db_sa().set(v)
+
+        # gives: assert [None, {'11': {'111': 42}}] == {'1': {'11': {'111': 42}}}
+        assert db_sa().get().val() == v
+
+
+class TestChildNavigation:
+    def test_get_child_none(self, db_sa):
+        assert db_sa().child('lorem').get().val() is None
+
+    def test_get_child_after_pushing_data(self, db_sa):
+        db_sa().set({'lorem': "a", 'ipsum': 2})
+
+        assert db_sa().child('lorem').get().val() == "a"
+        assert db_sa().child('ipsum').get().val() == 2
+
+    def test_update_child(self, db_sa):
+        db_sa().child('child').update({'c1/c11': 1, 'c1/c12': 2, 'c2': 3})
+
+        assert db_sa().child('child').child('c1').get().val() == {'c11': 1, 'c12': 2}
+        assert db_sa().child('child').child('c2').get().val() == 3
+
+    def test_path_equivalence(self, db_sa):
+        db_sa().set({'1': {'11': {'111': 42}}})
+
+        assert db_sa().child('1').child('11').child('111').get().val() == 42
+        assert db_sa().child('1/11/111').get().val() == 42
+        assert db_sa().child('1', '11', '111').get().val() == 42
+        assert db_sa().child(1, '11', '111').get().val() == 42
+
+
+class TestStreaming:
+    def test_create_stream_succeed(self, db_sa):
+        with make_append_stream(db_sa()) as (stream, l):
+            assert stream is not None
+
+    def test_does_initial_call(self, db_sa):
+        with make_append_stream(db_sa()) as (stream, l):
+            time.sleep(2)
+            assert len(l) == 1
+
+    def test_responds_to_update_calls(self, db_sa):
+        with make_append_stream(db_sa()) as (stream, l):
+            db_sa().set({"1": "a", "1_2": "b"})
+            db_sa().update({"2": "c"})
+            db_sa().push("3")
+
+            time.sleep(2)
+
+            assert len(l) == 3
```

Comment:
Gives `3` events 99% of the time, sometimes `4`.

The first event (data=None) is probably dropped by firebase, (`set` happens before the event is sent)
but this could also be a race condition relative to #70 (something like: the sse client is not listening when firebase sends the first event).